// rsync/file-browser.js
"use strict";

var fs = require('fs')
    , async = require('async')
    , path = require('path');

/**
 * @param path {String}
 * @param callback {function(String, RsyncNode)}
 */
var getNode = function (path, callback) {
    fs.stat(path, function (err, stats) {
        if (err) {
            callback(err, null);
            return;
        }
        var node = new RsyncNode();
        node.isDirectory = stats.isDirectory();
        node.path = path;
        node.bytes = stats.size;
        node.collapsed = true;
        node.children = undefined;
        callback(null, node);
    });
};

var RsyncNode = function RsyncNode() {
    this.isDirectory = undefined;
    this.path = undefined;
    this.bytes = undefined;
    this.collapsed = undefined;
    this.children = undefined;
};

/**
 * @param node {Node}
 * @param callback {function(String, Node)}
 */
var loadChildren = function (node, callback) {
    fs.readdir(node.path, function (err, files) {
        files.sort(function (a, b) {
            return a.toLowerCase().localeCompare(b.toLowerCase());
        });
        async.map(files, function (file, mapCallback) {
            getNode(path.join(node.path, file), mapCallback);
        }, function (error, results) {
            if (error) {
                callback(error);
            } else {
                node.children = results;
                callback(null, node);
            }
        })
    });
};

module.exports.getNode = getNode;
module.exports.loadChildren = loadChildren;
