// rsync/rsync-command/rsync-config.js
"use strict";

var RsyncConfig = function RsyncConfig() {
    this.rsyncConfigName = '';

    this.shortCommand = true;

    this.dryRun = true;
    this.delete = true;
    this.deleteAfter = true;
    this.filterFile = '';
    this.archive = true;
    this.oneFileSystem = true;
    this.logFile = '';
    this.itemizeChanges = true;
    this.verbose = true;
    this.humanReadable = true;
    this.progress = true;
    this.stats = true;

    this.src = '/tmp/rsync-test/src';
    this.dest = '/tmp/rsync-test/dest';
};

RsyncConfig.prototype.clone = function () {
    var clone = new RsyncConfig();
    clone.rsyncConfigName = this.rsyncConfigName;

    clone.shortCommand = this.shortCommand;

    clone.dryRun = this.dryRun;
    clone.delete = this.delete;
    clone.deleteAfter = this.deleteAfter;
    clone.filterFile = this.filterFile;
    clone.archive = this.archive;
    clone.oneFileSystem = this.oneFileSystem;
    clone.logFile = this.logFile;
    clone.itemizeChanges = this.itemizeChanges;
    clone.verbose = this.verbose;
    clone.humanReadable = this.humanReadable;
    clone.progress = this.progress;
    clone.stats = this.stats;

    clone.src = this.src;
    clone.dest = this.dest;
    return clone;
};

module.exports = RsyncConfig;
